### PR TITLE
Add Precompile `blake2_256`

### PR DESCRIPTION
**Pull request by AsceticBear**
After discussing with Aki, the `derive_account_id` sdk used needs this hash implementation.

image

### CODE REVIEW

File: node/runtime/pangolin/src/pallets/evm.rs [hunk 1]
```diff
@@ -138,9 +139,11 @@ where
 			>>::execute(input, target_gas, context, is_static)),
 			a if a == addr(25) =>
 				Some(<Dispatch<R>>::execute(input, target_gas, context, is_static)),
-			a if a == addr(26) => Some(<StateStorage<R, StorageFilter>>::execute(
+			a if a == addr(27) => Some(<StateStorage<R, StorageFilter>>::execute(
```
Comment (AsceticBear) [inline node/runtime/pangolin/src/pallets/evm.rs:142]: @wuminzhe I am going to update the `state-storage` precompile address of the pangolin network. See #1236 .

**Review by HackFisher — APPROVED** [review]

File: frame/dvm/evm/precompiles/blake2-256/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+authors     = ["Darwinia Network <hello@darwinia.network>"]
+description = "Blake2-256 hash precompiles for EVM pallet."
+edition     = "2021"
+homepage    = "https://darwinia.network"
+license     = "GPL-3.0"
+name        = "darwinia-evm-precompile-blake2-256"
+readme      = "README.md"
+repository  = "https://github.com/darwinia-network/darwinia-common"
+version     = "2.8.11"
+
```
Comment (AurevoirXavier) [inline frame/dvm/evm/precompiles/blake2-256/Cargo.toml:11]: An extra empty line.

File: frame/dvm/evm/precompiles/blake2-256/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Darwinia. If not, see <https://www.gnu.org/licenses/>.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+extern crate core;
```
Comment (AurevoirXavier) [inline frame/dvm/evm/precompiles/blake2-256/src/lib.rs:21]: Should we really need this?

**Comment by AurevoirXavier** [comment]
> image

It's always good to use a code block instead screenshot.

File: frame/dvm/evm/precompiles/blake2-256/Cargo.toml (hunk 2 above)
Comment (AsceticBear) [reply frame/dvm/evm/precompiles/blake2-256/Cargo.toml:11]: done

**Comment by AsceticBear** [comment]
> It's always good to use a code block instead screenshot.

Great suggestion :smile: I'll paste the code block next time.

**Comment by hujw77** [comment]
Blake2b precompile is `address(0x09)`

Refs: https://github.com/ethereum/EIPs/blob/master/EIPS/eip-152.md
https://github.com/paritytech/frontier/blob/master/frame/evm/precompile/blake2/src/lib.rs
https://github.com/darwinia-network/darwinia-bridges-sol/blob/master/contracts/utils/contracts/Blake2b.sol

@HackFisher @AsceticBear @wuminzhe

**Comment by AsceticBear** [comment]
This precompile is used to get the substrate [blake2b_256 hash](https://crates.parity.io/frame_support/struct.Blake2_256.html) result, which is different from the Blake2b you paste.

**Comment by hujw77** [comment]
I think it's the same one.

**Comment by AsceticBear** [comment]
> I think it's the same one.

This precompile requires the input must be fixed in length. https://github.com/paritytech/frontier/blob/master/frame/evm/precompile/blake2/src/lib.rs#L44-L50 but https://crates.parity.io/src/frame_support/hash.rs.html#159-161 has no such a restriction. Or there are some magics here?

**Comment by hujw77** [comment]
https://github.com/darwinia-network/darwinia-bridges-sol/blob/1d5afb2808af98d36994c08cdadd60bb8f6b463f/contracts/utils/contracts/Blake2b.sol#L93
Yes, 0xd5 match the length.

**Pull request merged**